src/dotnet/OpenAPI.Validator/Validation/ArmResourcePropertiesBag.cs

```diff
@@ -47,19 +47,54 @@ public class ArmResourcePropertiesBag : TypedRule<Dictionary<string, Schema>>
         {
-            var resModels = context.ResourceModels;
-            var violatingModels = resModels.Where(res => definitions[res].Properties?.ContainsKey("properties") == true
-                                                         && definitions[res].Properties["properties"]?.Properties?.Keys.Intersect(ArmPropertiesBag).Any() == true);
-            foreach (var violatingModel in violatingModels)
+            IEnumerable<string> resourceModelsWithPropertiesBag = context.ResourceModels.Where(resourceModel => definitions[resourceModel].Properties?.ContainsKey("properties") == true);
+            foreach(string resourceModelName in resourceModelsWithPropertiesBag)
             {
-                var violatingProperties = definitions[violatingModel].Properties.Keys.Union(definitions[violatingModel].Properties["properties"].Properties.Keys).Intersect(ArmPropertiesBag);
-                yield return new ValidationMessage(new FileObjectPath(context.File, context.Path.AppendProperty(violatingModel).AppendProperty("properties")), this, violatingModel,
-                                                   string.Join(", ", violatingProperties));
+                Dictionary<string, IEnumerable<string>> violations = new Dictionary<string, IEnumerable<string>>();             
+                CheckModelForViolation(definitions[resourceModelName], resourceModelName, definitions, context, violations);
+
+                foreach (string modelName in violations.Keys)
+                {
+                    yield return new ValidationMessage(new FileObjectPath(context.File,
+                    context.Path.AppendProperty(modelName).AppendProperty("properties")), this, modelName,
+                                                   string.Join(", ", violations[modelName]));
+                }
+            }
+        }
+
+        private string GetName(string reference)
+        {
+            return reference.Substring(reference.LastIndexOf("/") + 1);
+        }
+
+        private void CheckModelForViolation(Schema resourceModel, string resourceModelName, Dictionary<string, Schema> definitions, RuleContext context, Dictionary<string, IEnumerable<string>> violations)
+        {
+            if (resourceModel.Properties?.ContainsKey("properties") == true)
+            {
+                string referenceName = resourceModel.Properties["properties"].Reference;
+                if (referenceName != null)
+                {
+                    CheckModelForViolation(Schema.FindReferencedSchema(referenceName, definitions), GetName(referenceName), definitions, context, violations);
+                }
+
+                IEnumerable<string> violatingProperties = resourceModel.Properties["properties"].Properties?.Keys?.Intersect(ArmPropertiesBag);
+                if (violatingProperties?.Any() == true)
+                {
+                    violations[resourceModelName] = violatingProperties;
```

Comment:
Is it acceptable to overwrite the violations in case `resourceModelName` already existed in the `violations` dictionary?